src/PolyType/IMarshaler.cs

```diff
@@ -17,12 +19,14 @@ public interface IMarshaler<TSource, TTarget>
     /// </summary>
     /// <param name="value">The input of type <typeparamref name="TSource"/>.</param>
     /// <returns>A value of type <typeparamref name="TTarget"/> corresponding to <paramref name="value"/>.</returns>
+    [return: NotNullIfNotNull(nameof(value))]
```

Comment:
This is effectively equivalent to requiring that both input and output are non-nullable. The current annotation serves two purposes:

1. It enables mapping of null to non-null and vice-versa and
2. From the perspective of consuming library authors it doubles down on the notion that we cannot trust that implementations are honoring nullability annotations.

I think we could do better from the perspective of _authoring_ marshallers. One possibility is exposing an abstract `Marshaller<TSource, TTarget>` class that implements IMarshaller but uses different annotations for its methods and which short circuits null value handling.